cli/data.go

```diff
@@ -0,0 +1,195 @@
+		}
+		data := resp.GetData()
+		if len(data) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(data))
+		}
+
+		md := mds[0]
+		datum := data[0]
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return err
+		}
+		jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetId()+".json"))
+		if err != nil {
+			return err
+		}
+		if _, err := jsonFile.Write(mdJsonBytes); err != nil {
+			return err
+		}
+
+		// TODO: Include file name in metadata.
+
+		gzippedBytes := datum.GetBinary()
+		r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
+		if err != nil {
+			return err
+		}
+
+		// TODO: We need to store file extension too. In sync we map from ext -> mime type, so this is already available.
+		//       Or maybe we can just get this from file name.
+		ext, err := mimeTypeToFileExt(md.GetMimeType())
+		if err != nil {
+			return err
+		}
+
+		dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetId()+ext))
+		if _, err := io.Copy(dataFile, r); err != nil {
+			return err
+		}
+		if err := r.Close(); err != nil {
+			return err
+		}
+		skip++
+	}
+
+	return nil
+}
+
+func (c *AppClient) TabularData(dst string, filter *datapb.Filter) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	// Make dst/data and dst/metadata directory.
+	if err := os.MkdirAll(filepath.Join(dst, dataDir), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(dst, metadataDir), os.ModePerm); err != nil {
+		return err
+	}
+
+	resp, err := c.dataClient.TabularDataByFilter(context.Background(), &datapb.TabularDataByFilterRequest{
+		DataRequest: &datapb.DataRequest{
+			Filter: filter,
+			// TODO: For now don't worry about skip/limit. Just do everything in one request. Can implement batching when
+			//       tabular is implemented.
+		},
+		CountOnly: false,
+	})
+	if err != nil {
+		return err
+	}
+	mds := resp.GetMetadata()
+
+	for i, md := range mds {
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling metadata")
+		}
+		mdFile, err := os.Create(filepath.Join(dst, "metadata", strconv.Itoa(i)+".json"))
+		if err != nil {
+			return errors.Wrapf(err, "error creating metadata file %s", mdFile.Name())
+		}
+		if _, err := mdFile.Write(mdJsonBytes); err != nil {
+			return errors.Wrapf(err, "error writing metadata file %s", mdFile.Name())
+		}
+		if err := mdFile.Close(); err != nil {
+			return errors.Wrapf(err, "error closing metadata file %s", mdFile.Name())
+		}
+	}
+
+	data := resp.GetData()
+	// TODO: Use textpb insted of ndjson, and save multiple files.
+	dataFile, err := os.Create(filepath.Join(dst, "data", "data"+".ndjson"))
+	w := bufio.NewWriter(dataFile)
+	for _, datum := range data {
+		// Write everything as json for now.
+		d := datum.GetData()
+		if d == nil {
+			// TODO: This should never happen. Should this notify user? Error? Or just skip?
```

Comment:
If this should never happen, then it would seem like this should error.

Reply:
Yeah I was/am a bit conflicted here, because this should never happen, but it would actually indicate an error with the backend (that it stored invalid/empty data). Since it's not really an issue with the CLI, and the user has no ability to fix/avoid it (other than filing a bug report with us), I thought halting all the valid/working downloads could be more of an issue than the actual issue. I added the print mostly to help us debug - if we see that, we're doing something wrong on the backend.